languageServer/src/TextDocumentWrapper.ts

```diff
@@ -0,0 +1,294 @@
+
+        const workspaceEdit: WorkspaceEdit = {
+          changes: {
+            [this.uri]: textEdits
+          }
+        }
+
+        const codeAction = CodeAction.create(
+          'Add cmd files as dependencies',
+          workspaceEdit,
+          CodeActionKind.RefactorRewrite
+        )
+
+        codeActions.push(codeAction)
+      }
+    }
+
+    return codeActions
+  }
+
+  public getDefinitions(position: Position): Location[] {
+    const theSymbol = this.symbolAt(position)
+    if (theSymbol) {
+      const allDocs = this.documents.all()
+
+      const locationsAccumulator = []
+
+      for (const txtDoc of allDocs) {
+        const finder = this.createFinder(txtDoc)
+        const locations = finder.findLocationsFor(theSymbol)
+        locationsAccumulator.push(...locations)
+      }
+
+      return locationsAccumulator ?? []
+    }
+    return []
+  }
+
+  private getCompletionTemplates() {
+    const completionTemplates: CompletionTemplate[] = [
+      {
+        body: {
+          stages: { $1: { cmd: '$2' } }
+        },
+        label: 'stages'
+      },
+      {
+        body: {
+          $1: { cmd: '$2' }
+        },
+        label: 'Add stage'
+      }
+    ]
+    for (const path of this.pythonFilePaths) {
+      completionTemplates.push({
+        body: {
+          cmd: `python ${path} $1`
+        },
+        label: `cmd: ${path}`
+      })
+    }
+
+    return completionTemplates
+  }
+
+  private getTemplateExpressionSymbolsInsideScalar(
+    scalarValue: string,
+    nodeOffset: number
+  ) {
+    const templateSymbols: DocumentSymbol[] = []
+
+    const templates = scalarValue.matchAll(variableTemplates)
+    for (const template of templates) {
+      const expression = template[1]
+      const expressionOffset: number = nodeOffset + (template.index ?? 0) + 2 // To account for the '${'
+      const symbols = expression.matchAll(/[\dA-Za-z]+/g)
+      for (const templateSymbol of symbols) {
+        const symbolStart = (templateSymbol.index ?? 0) + expressionOffset
+        const symbolEnd = symbolStart + templateSymbol[0].length
+        const symbolRange = Range.create(
+          this.positionAt(symbolStart),
+          this.positionAt(symbolEnd)
+        )
+        templateSymbols.push(
+          DocumentSymbol.create(
+            templateSymbol[0],
+            undefined,
+            SymbolKind.Variable,
+            symbolRange,
+            symbolRange
+          )
+        )
+      }
+    }
+
+    return templateSymbols
+  }
+
+  private yamlScalarNodeToDocumentSymbols(
+    node: Scalar,
+    [nodeStart, valueEnd, nodeEnd]: [number, number, number]
+  ) {
+    const nodeValue = `${node.value}`
+
+    const symbolsSoFar: DocumentSymbol[] = [
+      DocumentSymbol.create(
+        nodeValue,
+        undefined,
+        SymbolKind.String,
+        Range.create(this.positionAt(nodeStart), this.positionAt(nodeEnd)),
+        Range.create(this.positionAt(nodeStart), this.positionAt(valueEnd))
```

Comment:
[Q] Why are the `range` and `selectionRange` not equal here but they are equal when you create `DocumentSymbols` in `getTemplateExpressionSymbolsInsideScalar`?